src/http_message.c

```diff
@@ -0,0 +1,134 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+#include "crt.h"
+
+#include "http.h"
+
+aws_crt_http_headers *aws_crt_http_headers_new_from_blob(const uint8_t *blob, size_t blob_length) {
+    aws_crt_http_headers *headers = aws_mem_calloc(aws_crt_allocator(), 1, sizeof(aws_crt_http_headers));
+    headers->headers = aws_http_headers_new(aws_crt_allocator());
+    struct aws_byte_cursor cursor = aws_byte_cursor_from_array(blob, blob_length);
+    while (cursor.len) {
+        uint32_t entry_len = 0;
+        /* Read header name: length|name */
+        if (!aws_byte_cursor_read_be32(&cursor, &entry_len)) {
+            goto bad_format;
+        }
+        struct aws_byte_cursor header_name = aws_byte_cursor_advance(&cursor, entry_len);
+
+        /* Read header value: length|value */
+        if (!aws_byte_cursor_read_be32(&cursor, &entry_len)) {
+            goto bad_format;
+        }
+        struct aws_byte_cursor header_value = aws_byte_cursor_advance(&cursor, entry_len);
+
+        struct aws_http_header header = {
+            .name = header_name,
+            .value = header_value,
+        };
+
+        aws_http_headers_add_header(headers->headers, &header);
+    }
+    return headers;
+
+bad_format:
+    aws_http_headers_release(headers->headers);
+    aws_mem_release(aws_crt_allocator(), headers);
+    aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+    return NULL;
+}
+
+void aws_crt_http_headers_acquire(aws_crt_http_headers *headers) {
+    aws_http_headers_acquire(headers->headers);
+}
+
+void aws_crt_http_headers_release(aws_crt_http_headers *headers) {
+    aws_http_headers_release(headers->headers);
+    aws_byte_buf_clean_up(&headers->encoded_headers);
+    aws_mem_release(aws_crt_allocator(), headers);
+}
+
+void aws_crt_http_headers_to_blob(aws_crt_http_headers *headers, uint8_t **out_blob, size_t *out_blob_length) {
+    aws_byte_buf_init(&headers->encoded_headers, aws_crt_allocator(), 256);
```

Comment:
shouldn't you check whether or not `encoded_headers` has already been filled out?